pkg/controller/dataimportcron-format-handlers.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2021 The CDI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+limitations under the License.
+See the License for the specific language governing permissions and
+*/
+
+package controller
+
+import (
+	"context"
+	"time"
+
+	snapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v6/apis/volumesnapshot/v1"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	"kubevirt.io/containerized-data-importer/pkg/common"
+	cc "kubevirt.io/containerized-data-importer/pkg/controller/common"
+)
+
+var formatHandlers = map[cdiv1.DataImportCronSourceFormat]formatHandler{
+	cdiv1.DataImportCronSourceFormatSnapshot: snapshotFormatHandler{},
+	cdiv1.DataImportCronSourceFormatPvc:      pvcFormatHandler{},
+}
+
+type formatHandler interface {
+	populateDataSourceSpec(*cdiv1.DataSourceSpec, *cdiv1.DataVolumeSourcePVC)
+	handleSourceCreation(context.Context, *DataImportCronReconciler, *cdiv1.DataImportCron, *cdiv1.DataVolume, *storagev1.StorageClass) error
```

Comment:
I am having a bit of a hard time with passing the reconciler into this function, I see we are using 4 things from the reconciler:
- 'Client'
- 'Log'
- 'setDataImportCronResourceLabels' (which really only cares about the `installerLabels`)
- 'deleteDvPvc' (which really only cares about the `Client`)
So we could relatively easily make deleteDvPvc not be part of the reconciler, then we would still have to pass at least 3 extra parameters to this function which also doesn't fill me with happiness.
Could we make some of these part of the implementing struct and not pass them as arguments to the function?

Reply:
That would make the initialization of these struct handlers not static anymore
instead, they would be initialized on every reconcile trigger which may also make us unhappy
https://github.com/kubevirt/containerized-data-importer/blob/295b5ca61245bbe11064a0780026fab83395ec2d/pkg/controller/dataimportcron-format-handlers.go#L35-L38